source/dispatch/semigroup_concat.js

```diff
@@ -0,0 +1,13 @@
+import _isArray from '../internal/_isArray';
+import _isObject from '../internal/_isObject';
+import _isString from '../internal/_isString';
+import array_concat from '../instances/array_concat';
+import object_concat from '../instances/object_concat';
+import string_concat from '../instances/string_concat';
+
+export default function semigroup_concat(a, b) {
+  if (_isArray(a)) return array_concat(a, b);
+  if (_isObject(a)) return object_concat(a, b);
+  if (_isString(a)) return string_concat(a, b);
+  throw new TypeError('No semigroup instance for ' + a);
+};
```

Comment:
This seems messy to me compared with the way mappings are defined in sanctuary-type-classes:

```javascript
var implementations = {
  Null: {
    prototype: {
      toString:                   Null$prototype$toString,
      'fantasy-land/equals':      Null$prototype$equals,
      'fantasy-land/lte':         Null$prototype$lte
    }
  },
  ...
};
```

Did you consider using a data structure rather than lots of branching logic?

Reply:
What do you mean by messy? Looks pretty clean to me? Obviously it is a bit boilerplate-y. I did think about a data structure but I think this is simple and naturally supports tree shaking with no overhead.

It would be neat if fantasy-land had a data file of some sort we could add as a dev dependency and use to auto-generate these files.

Reply:
> I did think about a data structure but I think this is simple and naturally supports tree shaking with no overhead.

Ah, I see. Your approach seems good, in that case. :thumbsup:

Reply:
I have a similar start at this, but also using a data structure.  This approach would definitely seem more modular.